### PR TITLE
Book A Doctors Appointment(Back-End)

### PR DESCRIPTION
**Pull request by Haywayaheadshot**
# What?

This is the back-end of an app that allows users to book doctor appointments. It was connected to [this](https://github.com/Haywayaheadshot/book-an-appointment-front-end) front-end repo.

# Why?

So the application can be easily maintained.

# How?
  - My coding partners and I;
      > Installed rails
      > Installed Ruby On Rails linters without the stylelint.
      > initialized the project backend with rails.
      > Created the backend to receive the frontend's signup data for user registration.
      > Updated the Cors file to include all methods
      > Added a table to handle username uniqueness so no two people can use the same username.
      > Created the User Login Page
      > Created the Api for the Login page
      > Created a doctors table.
      > Populated the doctors table.
      > Created a reservations table.
      > Created an API sending out all doctors.
      > Created an API to create and display reservations
      > Remove the views
      > Cleaned the code
      > Fix the routes
      > Removed commented lines in API controllers
      > Wrote test for the app

### Collaborators on this project are;
@eng-mohamed-eldeeb 
@Kelvin-Ben 
@asnganje 
@Haywayaheadshot

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LICENSE -->
+
+## 📝 License <a name="license"></a>
+
+This project is [MIT](./LICENSE) licensed.
+
+_NOTE: we recommend using the [MIT license](https://choosealicense.com/licenses/mit/) - you can set it up quickly by [using templates available on GitHub](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository). You can also use [any other license](https://choosealicense.com/licenses/) if you wish._
```
Comment (Meltrust) [inline README.md]: - [x] This is just a placeholder for you to create and include a license file.  Kindly remove this note, as it's meant just for you. 👍

File: README.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+Install this project with:
+
+- Run bundle install
+
+### Usage
+
+To run the project, execute the following command:
+
+[$ rails s]
```
Comment (Meltrust) [inline README.md]: - [x] Kindly, include in your instructions the required db commands that your users will need to run your backend app: `rails db:create, rails db:migrate, rails db:seed`. 👍   

- [x] I couldn't find any documentation for your API.  If it exists, kindly add instructions on your readme about how to access it, as this is a requirement for a team of 4 people.  If it does not exist yet, kindly implement it.👍

File: README.md [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+### Run tests
+
+To run tests, run the following command:
+
+
+Example command:
+
+```sh
+  bin/rails test test/models/article_test.rb
```
Comment (Meltrust) [inline README.md]: - [x] This command is not working on my end.  Also, I tried to run the tests found under the "test" folder but gave me errors.   Rspec worked fine (all 10 tests passed).

image

Kindly, correct this command in your instructions so all your tests pass.

- [ ] OPTIONAL:  If you are not using the "test" folder, kindly remove it

**Review by Meltrust — CHANGES_REQUESTED** [review]
_note: the [frontend review](https://github.com/Haywayaheadshot/book-an-appointment-front-end/pull/17) will be ready in a few minutes._

Hi @Haywayaheadshot,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Tests are passing✔️
-  Nice code organization ✔️ 
-  API is working well✔️ 
-  Good readme ✔️ 
-  Beautiful frontend ✔️ 



#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

**Comment by Haywayaheadshot** [comment]
@Meltrust Thank you for reviewing our project. We will work on the changes asap

**Review by Meltrust — APPROVED** [review]
_Note:  This review outcome applies **only** to the backend project.  Kindly, go to the frontend [PR](https://github.com/Haywayaheadshot/book-an-appointment-front-end/pull/17) to see the review comments for that project._

#### Hi @Haywayaheadshot,


# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause

### To highlight

- Readme has been vastly improved✔️ 
- Documentation has been added, and it looks great✔️ 
- Tests instructions have been updated✔️
- Awesome job ✔️  



Cheers and Happy coding!👏👏👏


------

**Comment by Haywayaheadshot** [comment]
@Meltrust Thank you for your kind words. Also, thank you for reviewing our project. It has been a great journey indeed. 🙏🏾🙏🏾

File: app/controllers/api/doctors_controller.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def destroy
+    @doctor = Doctor.find_by(id: params[:id])
+    if @doctor
+      @doctor.transaction do
+        @doctors_reservations = DoctorsReservation.where(doctor_id: @doctor.id)
+        @reservations = Reservation.where(id: @doctors_reservations.map(&:reservation_id))
+        @reservations.destroy_all # deletes all reservations of the doctor
+        @doctors_reservations.destroy_all # deletes all records linking the doctor and reservations
```
Comment (Meltrust) [inline app/controllers/api/doctors_controller.rb:15]: - [ ] Kindly, remove this old commented code in order to improve the repo's cleanliness. 👍

**Pull request merged**